xds/internal/xdsclient/authority.go

```diff
@@ -86,89 +98,135 @@ type authority struct {
-// new authority. All fields here correspond directly to appropriate fields
-// stored in the authority struct.
-type authorityArgs struct {
-	// The reason for passing server config and bootstrap config separately
-	// (although the former is part of the latter) is because authorities in the
-	// bootstrap config might contain an empty server config, and in this case,
-	// the top-level server config is to be used.
-	serverCfg          *bootstrap.ServerConfig
-	bootstrapCfg       *bootstrap.Config
-	serializer         *grpcsync.CallbackSerializer
-	resourceTypeGetter func(string) xdsresource.Type
-	watchExpiryTimeout time.Duration
-	backoff            func(int) time.Duration // Backoff for ADS and LRS stream failures.
-	logger             *grpclog.PrefixLogger
+// authorityBuildOptions wraps arguments required to create a new authority.
+type authorityBuildOptions struct {
+	serverConfigs    bootstrap.ServerConfigs      // Server configs for the authority
+	name             string                       // Name of the authority
+	serializer       *grpcsync.CallbackSerializer // Callback serializer for invoking watch callbacks
+	getChannelForADS xdsChannelForADS             // Function to acquire a reference to an xdsChannel
+	logPrefix        string                       // Prefix for logging
 }
 
-func newAuthority(args authorityArgs) (*authority, error) {
+// newAuthority creates a new authority instance with the provided
+// configuration. The authority is responsible for managing the state of
+// resources requested from the management server, as well as acquiring and
+// releasing references to channels used to communicate with the management
+// server.
+//
+// Note that no channels to management servers are created at this time. Instead
+// a channel to the first server configuration is created when the first watch
+// is registered, and more channels are created as needed by the fallback logic.
+func newAuthority(args authorityBuildOptions) *authority {
+	ctx, cancel := context.WithCancel(context.Background())
 	ret := &authority{
-		serverCfg:          args.serverCfg,
-		bootstrapCfg:       args.bootstrapCfg,
-		serializer:         args.serializer,
-		resourceTypeGetter: args.resourceTypeGetter,
-		watchExpiryTimeout: args.watchExpiryTimeout,
-		logger:             args.logger,
-		resources:          make(map[xdsresource.Type]map[string]*resourceState),
+		name:                      args.name,
+		watcherCallbackSerializer: args.serializer,
+		getChannelForADS:          args.getChannelForADS,
+		xdsClientSerializer:       grpcsync.NewCallbackSerializer(ctx),
+		xdsClientSerializerClose:  cancel,
+		resources:                 make(map[xdsresource.Type]map[string]*resourceState),
+	}
+
+	// Create a prefix logger specific to this authority.
+	l := grpclog.Component("xds")
+	logPrefix := args.logPrefix + fmt.Sprintf("[authority %q] ", args.name)
+	ret.logger = igrpclog.NewPrefixLogger(l, logPrefix)
```

Comment:
Optional: could declare l and logPrefix before ret and inline ret.logger creation.

Reply:
Yeah I made this comment on other ones which didn't work because of that prefix thing haha so this time I made sure it didn't require the struct.